CMakeLists.txt

```diff
@@ -0,0 +1,428 @@
+option(FREERTOS_PORT_IAR_STR75X "FreeRTOS port (IAR) for STR75x")
+option(FREERTOS_PORT_IAR_STR91X "FreeRTOS port (IAR) for STR91x")
+option(FREERTOS_PORT_IAR_V850ES_FX3 "FreeRTOS port (IAR) for Renesas V850ES/Fx3")
+option(FREERTOS_PORT_IAR_V850ES_HX3 "FreeRTOS port (IAR) for Renesas V850ES/Hx3")
+option(FREERTOS_PORT_MIKROC_ARM_CM4F "FreeRTOS port (MikroC) for ARM Cortex-M4 with FPU")
+option(FREERTOS_PORT_MPLAB_PIC18F "FreeRTOS port (MPLAB) for PIC18F")
+option(FREERTOS_PORT_MPLAB_PIC24 "FreeRTOS port (MPLAB) for PIC24")
+option(FREERTOS_PORT_MPLAB_PIC32MEC14XX "FreeRTOS port (MPLAB) for PIC32MEC14xx")
+option(FREERTOS_PORT_MPLAB_PIC32MX "FreeRTOS port (MPLAB) for PIC32MX")
+option(FREERTOS_PORT_MPLAB_PIC32MZ "FreeRTOS port (MPLAB) for PIC32MZ")
+option(FREERTOS_PORT_MSVC_MINGW "FreeRTOS port for x86 MSVC or MinGW")
+option(FREERTOS_PORT_OWATCOM_16BIT_DOS_FLSH186 "FreeRTOS port (Open Watcom) for 16 bit DOS Flsh186")
+option(FREERTOS_PORT_OWATCOM_16BIT_DOS_PC "FreeRTOS port (Open Watcom) for 16 bit DOS PC")
+option(FREERTOS_PORT_PARADIGM_TERN_EE_LARGE "FreeRTOS port (Paradigm) for Tern EE large")
+option(FREERTOS_PORT_PARADIGM_TERN_EE_SMALL "FreeRTOS port (Paradigm) for Tern EE small")
+option(FREERTOS_PORT_RENESAS_RX100 "FreeRTOS port (Renesas) for RX100")
+option(FREERTOS_PORT_RENESAS_RX200 "FreeRTOS port (Renesas) for RX200")
+option(FREERTOS_PORT_RENESAS_RX600 "FreeRTOS port (Renesas) for RX600")
+option(FREERTOS_PORT_RENESAS_RX600_V2 "FreeRTOS port (Renesas) for RX600 v2")
+option(FREERTOS_PORT_RENESAS_RX700_V3_DPFPU "FreeRTOS port (Renesas) for RX700 v3 with DPFPU")
+option(FREERTOS_PORT_RENESAS_SH2A_FPU "FreeRTOS port (Renesas) for SH2A with FPU")
+option(FREERTOS_PORT_ROWLEY_MSP430F449 "FreeRTOS port (Rowley) for MSP430F449")
+option(FREERTOS_PORT_RVDS_ARM_CA9 "FreeRTOS port (RVDS) for ARM Cortex-A9")
+option(FREERTOS_PORT_RVDS_ARM_CM0 "FreeRTOS port (RVDS) for ARM Cortex-M0")
+option(FREERTOS_PORT_RVDS_ARM_CM3 "FreeRTOS port (RVDS) for ARM Cortex-M3")
+option(FREERTOS_PORT_RVDS_ARM_CM4_MPU "FreeRTOS port (RVDS) for ARM Cortex-M4 with MPU")
+option(FREERTOS_PORT_RVDS_ARM_CM4F "FreeRTOS port (RVDS) for ARM Cortex-M4 with FPU")
+option(FREERTOS_PORT_RVDS_ARM_CM7 "FreeRTOS port (RVDS) for ARM Cortex-M7")
+option(FREERTOS_PORT_RVDS_ARM7_LPC21XX "FreeRTOS port (RVDS) for ARM7 LPC21xx")
+option(FREERTOS_PORT_SDCC_CYGNAL "FreeRTOS port (SDCC) for Cygnal")
+option(FREERTOS_PORT_SOFTUNE_MB91460 "FreeRTOS port (Softune) for MB91460")
+option(FREERTOS_PORT_SOFTUNE_MB96340 "FreeRTOS port (Softune) for MB96340")
+option(FREERTOS_PORT_TASKING_ARM_CM4F "FreeRTOS port (Tasking) for ARM Cortex-M4 with FPU")
+option(FREERTOS_PORT_CDK_THEAD_CK802 "FreeRTOS port (CDK) for T-head CK802")
+option(FREERTOS_PORT_XCC_XTENSA "FreeRTOS port (XCC) for Xtensa")
+option(FREERTOS_PORT_WIZC_PIC18 "FreeRTOS port (WizC) for PIC18")
+
+add_library(freertos STATIC
+    croutine.c
+    event_groups.c
+    list.c
```

Comment:
Will this method of including files work if the CMake is invoked from other directories?
Maybe we should add the full path of FreeRTOS-Kernel before the file names like so:

```
# Set global path variables.
get_filename_component( __MODULE_ROOT_DIR "${CMAKE_CURRENT_LIST_DIR}" ABSOLUTE )
set( MODULE_ROOT_DIR ${__MODULE_ROOT_DIR} CACHE INTERNAL "FreeRTOS-Kernel repository root." )

add_library(freertos STATIC
            ${MODULE_ROOT_DIR}/list.c
            .
            .
            ) 
```

Reply:
CMake modules are intended to be integrated into the top-level project using `add_subdirectory(third_party/FreeRTOS-Kernel)` and not `include(third_party/FreeRTOS-Kernel/CMakeLists.txt)`

So, using relative paths to sources is how it works with add_subdirectory(...)

Reply:
I understand that, however, I was concerned about what happens when someone just tries to compile the Kernel using Cmake from some other directory. Something like this:
```
cmake -S ./Test/FreeRTOS-Kernel -b build
```

Would that throw an error? I will try it out locally :)

Reply:
Here is what I get in such case:
```cmd
cmake -Bbuild -G Ninja -S third_party/FreeRTOS-Kernel -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ && cmake --build build
```
output:
```
CMake Error at CMakeLists.txt:8 (message):
   FreeRTOSConfig.h file directory not specified. Please specify absolute path to it from top-level CMake file:
    set(FREERTOS_CONFIG_FILE_DIRECTORY <absolute path to FreeRTOSConfig.h directory> CACHE STRING "")
   or from CMake command line option:
    -DFREERTOS_CONFIG_FILE_DIRECTORY='/absolute_path/to/FreeRTOSConfig.h/directory'

-- Configuring incomplete, errors occurred!
```

But if I set `FREERTOS_CONFIG_FILE_DIRECTORY` and `FREERTOS_PORT` the build succeeds:
```
cmake -Bbuild -G Ninja -S third_party/FreeRTOS-Kernel -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DFREERTOS_PORT=MSVC_MINGW -DFREERTOS_CONFIG_FILE_DIRECTORY=D:\dev\embedded\projects\mcu-framework\stm32f4 && cmake --build build

CMake Warning (dev) in CMakeLists.txt:
  No project() command is present.  The top-level CMakeLists.txt file must
  contain a literal, direct call to the project() command.  Add a line of
  code such as

    project(ProjectName)

  near the top of the file, but after cmake_minimum_required().

  CMake is pretending there is a "project(Project)" command on the first
  line.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/mydevtools/mingw64/bin/gcc.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/mydevtools/mingw64/bin/g++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: D:/dev/embedded/projects/mcu-framework/stm32f4/build
[10/11] Building C object CMakeFiles/freertos.dir/portable/Common/mpu_wrappers.c.obj
D:/dev/embedded/projects/mcu-framework/stm32f4/third_party/FreeRTOS-Kernel/portable/Common/mpu_wrappers.c: In function 'MPU_xTaskCreate':
D:/dev/embedded/projects/mcu-framework/stm32f4/third_party/FreeRTOS-Kernel/portable/Common/mpu_wrappers.c:61:9: warning: implicit declaration of function 'xPortRaisePrivilege' [-Wimplicit-function-declaration]
   61 |         xPortRaisePrivilege( xRunningPrivileged );
      |         ^~~~~~~~~~~~~~~~~~~
D:/dev/embedded/projects/mcu-framework/stm32f4/third_party/FreeRTOS-Kernel/portable/Common/mpu_wrappers.c:63:9: warning: implicit declaration of function 'vPortResetPrivilege' [-Wimplicit-function-declaration]
   63 |         vPortResetPrivilege( xRunningPrivileged );
      |         ^~~~~~~~~~~~~~~~~~~
[11/11] Linking C static library libfreertos_kernel.a
```

Reply:
Awesome! Thanks for testing that. It puts my concern to rest! :)